### ISSUE TITLE
NaN cost with multi-source transformer

### ISSUE THREAD

**Issue by okgrammer** [marian-nmt/marian-dev#570]
Hello, I am trying out the multi-source translation in Marian. I am using two source input that consists of 1) the source sentence and 2) additional annotation information, such as certain phrase span in this sentence. The second input sequence is typically really short. I keep getting NaN cost value by training in this way, even after changing the workspace to a small size and reducing warmup steps. Here is my training command for reference (using Marian v1.7.8):

```
./build/marian --model $model_dir/model.npz --type multi-transformer \
--train-sets train.src1 train.src2 train.trg \
--vocabs vocab.spm vocab.spm vocab.spm \
--dim-vocabs 32000 32000 32000 \
--mini-batch-fit --mini-batch 1000 --maxi-batch 1000 \
--valid-freq 5000 --save-freq 5000 --disp-freq 500 \
--valid-metrics ce-mean-words perplexity bleu-detok \
--valid-sets valid.src1 valid.src2 valid.trg \
--quiet-translation \
--beam-size 6 --normalize=0.6 \
--valid-mini-batch 16 \
--early-stopping 5 --cost-type=ce-mean-words \
--log $model_dir/train.log --valid-log $model_dir/valid.log \
--enc-depth 6 --dec-depth 6 \
--transformer-preprocess n --transformer-postprocess da \
--tied-embeddings-all --dim-emb 1024 --transformer-dim-ffn 4096 \
--transformer-dropout 0.1 --transformer-dropout-attention 0.1 --transformer-dropout-ffn 0.1 --label-smoothing 0.1 \
--learn-rate 0.0001 --lr-warmup 4000 --lr-decay-inv-sqrt 4000 --lr-report \
--optimizer-params 0.9 0.98 1e-09 --clip-norm 5 \
--devices 0 1 2 3 --workspace 5000 --sync-sgd --seed 1 \
--exponential-smoothing \
--shuffle-in-ram
```

**Comment by snukky** [comment]
Are you getting NaN costs from the beginning of training or after some batches are processed? (`--disp-first N` will display first N updates)
If it's because of instability of the training, I would actually try the opposite: elongating the warmup, increasing the batch size and maybe decreasing dropout.

**Comment by okgrammer** [comment]
The first update printed a cost of about 500, then it became NaN on the second update onwards. I tried with warmup steps up to 16k and maximizing workspace to about 27GB, but still the same NaN cost was produced. It is able to train after reducing the number of layers 5 though.

**Comment by emjotde** [comment]
What GPU is that?

**Comment by emjotde** [comment]
Also try this, I think these are more stable:

```
--transformer-preprocess "" --transformer-postprocess dan \
--tied-embeddings-all --dim-emb 1024 --transformer-dim-ffn 4096 \
--transformer-dropout 0.2 --transformer-dropout-attention 0.0 --transformer-dropout-ffn 0.1 --label-smoothing 0.1 \
--learn-rate 0.0001 --lr-warmup 4000 --lr-decay-inv-sqrt 4000 --lr-report \
--optimizer-params 0.9 0.98 1e-09 0.01 --clip-norm 0 \
```

**Comment by okgrammer** [comment]
Here's the output from `--version`: `v1.7.8 4d57f87a 2019-03-01 18:41:40 +0000`. I tried it on 32GB V100 GPUs. BTW, it worked with 6 layers and the above settings you suggested and training is stable. Thanks for your help!

**Comment by alvations** [comment]
Just curious, what's the intuition behind:

 - `--transformer-preprocess "" --transformer-postprocess dan` as opposed to `--transformer-preprocess n --transformer-postprocess da`, essentially, they are the same right? Just that only the first transformer block don't get normalized

 - `--clip-norm 0`: Disable normalization clipping because ?? 

- `--transformer-dropout 0.2 --transformer-dropout-attention 0.0` is there little to none attention dropout because it's a student model? Any intuition why `--transformer-dropout 0.1 --transformer-dropout-attention 0.1` would be less appropriate?

 - `--optimizer-params ...`: By default there are three numbers there, https://github.com/marian-nmt/marian-dev/blob/master/src/common/aliases.cpp#L102 , can we assume, it's `beta1`, `beta2` and `epsilon`? Then what is the 4th parameter? Decay from https://github.com/marian-nmt/marian-dev/blob/master/src/optimizers/optimizers.h#L195 ? Any reason to decay the adam as oppose to normal adam?


Sorry for the multiple questions all at once, just very curious about the hyperparameter choices =)

**Comment by emjotde** [comment]
@okgrammer Great that this worked. If you are working with `marian-dev` anyway I recommend to update to current master, we are at version 1.8.33 already, bascially ready to release 1.9. Quite a few changes there might impact stability.

**Comment by emjotde** [comment]
@alvations Ok. Let's see if I can turn my brains hyperparameter optimizer into words :)

1) `--transformer-preprocess "" --transformer-postprocess dan` is normalizing the _last_ layer, I found that this behaves a bit more sane, especially with deeper models. Many say the opposite, but I could never confirm this. The multi-encoder model is a bit deeper, so I would go with that. 
2) Clipping, because it's essentially broken now and over-clipping. A while ago we changed to summing costs instead of averaging and have forgotten to adapt that. I am working on that. Usually it won't affect you, but since we were seeing NaNs here I asked to disable.
3) No dropout in attention: Again I found dropout there to make problems with training stability especially when using `--transformer-preprocess n --transformer-postprocess da`. The dropout zeros out whole words and then with the skip connection the zeros arrive at the top. That seems to drive up intial cost quite a bit, becomes worse when the model is deeper. 
4) Yes, that's weight decay. NaNs usually happen when something weird happens in the gradients, usually overflow. You want to bring your gradient norm down, gradient norm correlation with parameter norm, weight decay keeps your parameters small.

**Comment by emjotde** [comment]
@okgrammer also, if you see more problems with NaNs that is likely caused by TensorCore math on the V100. This is doing by default 16bit matrix-multiplication, that can overflow in the gradients. One (undocumented) hack to turn off tensor cores in Marian is to run like this:

```
export ENABLE_CUBLAS_TENSOR_OP_MATH_FP32=0
./marian ...
```

**Comment by okgrammer** [comment]
Thanks @emjotde, I also experimented by only changing to `--transformer-preprocess "" --transformer-postprocess dan` and keeping the rest of the settings unchanged from my initial post. It turns out that this alone can also prevent the NaN cost from happening. Another observation is that the loss decreases more rapidly: cost is ~10 after 100 updates, typically the cost is about 200 or more when using `--transformer-preprocess n`. And yes will update to the latest version soon. :)

Additional question, would it be possible to use a different configuration for each encoder? For example, 6 layers for the first encoder but only 5 or less for the second encoder.

**Comment by emjotde** [comment]
As for the question, not without code changes. BTW, There is a model that is sharing parameters between encoders, I always found that to be superior: `--type shared-multi-transformer`

**Comment by emjotde** [comment]
Closing. Op seems happy and the NaNs is not something we can easily solve if this is indeed caused by overflow due to tensorcores. The will be a few more NaN guards available in the future, already working on that.
